contracts/DegenLockToken.sol

```diff
@@ -0,0 +1,185 @@
+// SPDX-License-Identifier: LGPL-3.0-only
+pragma solidity ^0.8.0;
+
+import {IERC20, SafeERC20} from "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+import {ERC20} from "@openzeppelin/contracts/token/ERC20/ERC20.sol";
+import {Ownable} from "@openzeppelin/contracts/access/Ownable.sol";
+import {ReentrancyGuard} from "@openzeppelin/contracts/utils/ReentrancyGuard.sol";
+
+/**
+ * @notice Lock Degen for a set amount of time.
+ * @custom:security-contact jacek@degen.tips
+ */
+contract DegenLockToken is ERC20, Ownable, ReentrancyGuard {
+    using SafeERC20 for IERC20;
+
+    mapping(address account => uint256) public depositTimestamps;
+
+    /**
+     * @dev Name of the token representing the claim on the locked token
+     */
+    string public constant TOKEN_NAME = "Locked Degen";
+
+    /**
+     * @dev Symbol of the token representing the claim on the locked token
+     */
+    string public constant TOKEN_SYMBOL = "LDEGEN";
+
+    /**
+     * @dev The ERC20 token to be locked
+     */
+    IERC20 public constant TOKEN =
+        IERC20(0xfee293840D23B0B2De8C55e1Cf7A9F01C157767c);
+
+    /**
+     * @dev Lock duration in seconds, period starts after the deposit timestamp
+     */
+    uint256 public lockDuration = 90 days;
+
+    /**
+     * @dev Minimum amount of tokens that have to be deposited
+     */
+    uint256 public minDepositAmount = 1e22;
+
+    /**
+     * @dev Withdraw is not possible because the lock period is not over yet
+     */
+    error LockPeriodOngoing();
+
+    /**
+     * @dev ERC-20 function is not supported
+     */
+    error NotSupported();
+
+    /**
+     *  @dev The lock duration is longer than a year
+     */
+    error LockDurationIsTooLong();
+
+    /**
+     *  @dev Amount of tokens to be locked is zero
+     */
+    error ZeroAmount();
+
+    /**
+    /**
+     * @dev Minimum deposit not met
+     */
+    error MinimumDepositNotMet();
+
+    /**
+     *  @dev The lock duration has been updated
+     */
+    event LockDurationUpdated(uint256 duration);
+
+    /**
+     *  @dev The minimum deposit amount has been updated
+     */
+    event MinDepositAmountUpdated(uint256 minDepositAmount);
+
+    /**
+     *  @dev The deposit timestamp has been set
+     */
+    event DepositTimestampUpdated(
+        address indexed account,
+        uint256 depositTimestamp
+    );
+
+    /**
+     * @dev Construct a new Degen token
+     */
+    constructor() ERC20(TOKEN_NAME, TOKEN_SYMBOL) Ownable(msg.sender) {}
+
+    /**
+     * @dev Deposit tokens to be locked until the end of the locking period
+     * @param amount The amount of tokens to deposit
+     */
+    function deposit(uint256 amount) external nonReentrant {
+        if (amount < minDepositAmount) {
+            revert MinimumDepositNotMet();
+        }
+
+        TOKEN.safeTransferFrom(msg.sender, address(this), amount);
+        _mint(msg.sender, amount);
+
+        depositTimestamps[msg.sender] = block.timestamp;
```

Comment:
I agree with this recommendation.
Every deposit should manage separately.

Reply:
I think we'll make it so that the unlock duration can only be reduced. If we need vaults with a 1-year lock in the future, we can create a separate 1-year lock contract. This will be easier to manage than having some users with one lock duration and others with a different one in the same contract.

Reply:
@jacek0x 
Even if there is separate vault some user might have an issue.
Think about user want stake 10000 DEGEN for 1 year. If user want to increase amount after stake, unlock time will extend for 1 year.

1. Alice stake 10000 DEGEN at 24.07.05.
2. Alice want stake more DEGEN at 24.12.01. -> Alice's unlock time reset to 25.12.01.
3. Alice try unstake 10000 DEGEN at 25.07.05. but revert

This inconvenience can make users less willing to stake more.

Reply:
Yes, the unlock time extends when you add more funds, and it’s a UX issue. If you don’t want to extend your lock period, using a different wallet for the new stake might be easier. It keeps things simple, which is important, especially for the first version. We can always improve later, but I personally like the default lock extension mechanism.